engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/job/preparation/SubWorkflowPreparationActor.scala

```diff
@@ -20,24 +24,48 @@ class SubWorkflowPreparationActor(workflowDescriptor: EngineWorkflowDescriptor,
 
   lazy val workflowIdForLogging = workflowDescriptor.id
 
-  def prepareExecutionActor(inputEvaluation: WomEvaluatedCallInputs): CallPreparationActorResponse = {
+  def prepareExecutionActor(inputEvaluation: WomEvaluatedCallInputs): ErrorOr[CallPreparationActorResponse] = {
     val oldBackendDescriptor = workflowDescriptor.backendDescriptor
 
-    val newBackendDescriptor = oldBackendDescriptor.copy(
-      id = subWorkflowId,
-      callable = callKey.node.callable,
-      // TODO WOM: need FQN for input definitions somehow ? For now don't worry about subWF
-      knownValues = Map.empty,//workflowDescriptor.knownValues ++ (inputEvaluation map { case (k, v) => k.fullyQualifiedName -> v }),
-      breadCrumbs = oldBackendDescriptor.breadCrumbs :+ BackendJobBreadCrumb(workflowDescriptor.callable, workflowDescriptor.id, callKey)
-    )
-    val engineDescriptor = workflowDescriptor.copy(backendDescriptor = newBackendDescriptor, parentWorkflow = Option(workflowDescriptor))
-    SubWorkflowPreparationSucceeded(engineDescriptor, inputEvaluation)
+    evaluateStartingKnownValues(inputEvaluation, callKey.node.callable.graph.inputNodes) map { startingValues =>
+      val newBackendDescriptor = oldBackendDescriptor.copy(
+        id = subWorkflowId,
+        callable = callKey.node.callable,
+        knownValues = startingValues,
+        breadCrumbs = oldBackendDescriptor.breadCrumbs :+ BackendJobBreadCrumb(workflowDescriptor.callable, workflowDescriptor.id, callKey)
+      )
+      val engineDescriptor = workflowDescriptor.copy(backendDescriptor = newBackendDescriptor, parentWorkflow = Option(workflowDescriptor))
+      SubWorkflowPreparationSucceeded(engineDescriptor, inputEvaluation)
+    }
+  }
+
+  def evaluateStartingKnownValues(inputEvaluation: WomEvaluatedCallInputs, workflowInputs: Set[GraphInputNode]): ErrorOr[Map[OutputPort, WomValue]] = {
+    val providedInputs: Map[OutputPort, WomValue] = inputEvaluation map { case (name, value) =>
+      workflowInputs.collectFirst{
+        case i if i.singleOutputPort.identifier.localName.value == name.localName.value => i.singleOutputPort
+      } -> value
+    } collect {
+      case (Some(k), v) => k -> v
+    }
+
+    // We don't need to supply these:
+    val optionalsAndDefaults: Set[GraphNode] = workflowInputs collect {
+      case optionalInputNode: OptionalGraphInputNode => optionalInputNode
+      case optionalWithDefault: OptionalGraphInputNodeWithDefault => optionalWithDefault
+    }
+
+    // Make sure we got everything:
```

Comment:
Question to reviewers: should we also make sure we didn't get anything we didn't want?

Reply:
Good question. In general, I'm satisfied if a unit test makes sure we didn't get anything we didn't want, instead of production code testing/asserting during real workflow runs.

However, at the moment, the codecov chrome plugin says this file has _zero_ lines covered.

Reply:
This is more asking “should we (the engine) check that the ‘{ input: x=... }‘ block didn’t give us inputs that we (the sub workflow) weren’t expecting (albeit, perhaps we should have caught that before subworkflow preparation...)

Reply:
I'm still digesting all of this code, so apologies if I'm missing something obvious.

What are reasons why the engine would/should check for extra inputs? If the check would just be for ensuring that the code is correct, then I'd say put sanity checks in test code, not in main code. If a bug happens, I'd put the check into a regression test.

The one case I have seen main-code-asserts is something like (pseudo-code):

```
if (logger.isDebugModeEnabled) {
   val extraCheckResults = doExtraCheckHere()
   extraCheckResults foreach logger.debug
}
```

But those blocks I've usually reserved for cases where some sort of transient/heisen bug was happening in production environments. When discovered I usually replaced the blocks with a proper regression test.

Reply:
I *think* the question is if a *user* has written WDL where the main workflow calls a subworkflow with inputs that aren't valid for that subworkflow, should Cromwell call that out (where "call that out" is left ambiguous as to warning or error)? Right now I believe this condition is silently ignored.

Reply:
Right, @mcovarr!

Hopefully by this point something else would have checked statically so this could be a sanity check, but OTOH I don’t think that “something else” exists yet...

Reply:
That's a WDL spec decision for the committee to be debated elsewhere outside of this particular engine.

That said, and invoking "ToL" ;)

ToL: I'm fine with (and assumed that) an engine would ignore extra inputs to a sub workflow. Only missing inputs are an error. This is based on the behavior of the inputs.json, that I believe _does not_ cause an error if there are extra fields not found in the workflow. I'm not sure if that's in the "WDL Spec" either? Honestly I'm fine with whatever, but I'd like consistency.

Reply:
The spec currently isn't too particular on this topic: https://github.com/openwdl/wdl/blob/master/versions/draft-3/SPEC.md#specifying-workflow-inputs-in-json

This means that until someone wants to try to make it part of the spec, it isn't.